File: ChessProject-PHP/tests/ChessBoardTest.php
<?php

namespace SolarWinds\Chess;

use SolarWinds\Chess\ChessBoard;
use SolarWinds\Chess\PieceColorEnum;
use SolarWinds\Chess\Pawn;

class ChessBoardTest extends \PHPUnit\Framework\TestCase
{

    /** @var  ChessBoard */
    private $_testSubject;

    public function setUp(): void
    {
        $this->_testSubject = new ChessBoard();
    }

    public function testHas_MaxBoardWidth_of_7()
    {
        $this->assertEquals(7, ChessBoard::MAX_BOARD_WIDTH);
    }

    public function testHas_MaxBoardHeight_of_7()
    {
        $this->assertEquals(7, ChessBoard::MAX_BOARD_HEIGHT);
    }

    public function testIsLegalBoardPosition_True_X_equals_0_Y_equals_0()
    {
        $isValidPosition = $this->_testSubject->isLegalBoardPosition(0, 0);
        $this->assertTrue($isValidPosition);
    }

    public function testIsLegalBoardPosition_True_X_equals_5_Y_equals_5()
    {
        $isValidPosition = $this->_testSubject->isLegalBoardPosition(5, 5);
        $this->assertTrue($isValidPosition);
    }

    public function testIsLegalBoardPosition_False_X_equals_11_Y_equals_5()
    {
        $isValidPosition = $this->_testSubject->isLegalBoardPosition(11, 5);
        $this->assertFalse($isValidPosition);
    }

    public function testIsLegalBoardPosition_False_X_equals_0_Y_equals_9()
    {
        $isValidPosition = $this->_testSubject->isLegalBoardPosition(0, 9);
        $this->assertFalse($isValidPosition);
    }

    public function testIsLegalBoardPosition_False_X_equals_11_Y_equals_0()
    {
        $isValidPosition = $this->_testSubject->isLegalBoardPosition(11, 0);
        $this->assertFalse($isValidPosition);
    }

    public function testIsLegalBoardPosition_False_For_Negative_Y_Values()
    {
        $isValidPosition = $this->_testSubject->isLegalBoardPosition(5, -1);
        $this->assertFalse($isValidPosition);
    }

    public function testAvoids_Duplicate_Positioning()
    {
        $firstPawn = new Pawn(PieceColorEnum::BLACK());
        $secondPawn = new Pawn(PieceColorEnum::BLACK());
        $this->_testSubject->add($firstPawn, 6, 3, PieceColorEnum::BLACK());
        $this->_testSubject->add($secondPawn, 6, 3, PieceColorEnum::BLACK());
        $this->assertEquals(6, $firstPawn->getXCoordinate());
        $this->assertEquals(3, $firstPawn->getYCoordinate());
        $this->assertEquals(-1, $secondPawn->getXCoordinate());
        $this->assertEquals(-1, $secondPawn->getYCoordinate());
    }

    public function testLimits_The_Number_Of_Pawns()
    {
        for ($i = 0; $i < 10; $i++) {
            $pawn = new Pawn(PieceColorEnum::BLACK());
            $row = $i / ChessBoard::MAX_BOARD_WIDTH;
            $this->_testSubject->add($pawn, 6 + $row, $i % ChessBoard::MAX_BOARD_WIDTH, PieceColorEnum::BLACK());
            if ($row < 1) {
                $this->assertEquals(6 + $row, $pawn->getXCoordinate());
                $this->assertEquals($i % ChessBoard::MAX_BOARD_WIDTH, $pawn->getYCoordinate());
            } else {
                $this->assertEquals(-1, $pawn->getXCoordinate());
                $this->assertEquals(-1, $pawn->getYCoordinate());
            }
        }
    }

}

File: ChessProject-Csharp/tests/ChessBoardTest.cs
using NUnit.Framework;

namespace SolarWinds.MSP.Chess
{
    [TestFixture]
    public class ChessBoardTest
    {
        private ChessBoard chessBoard;

        [SetUp]
        public void SetUp()
        {
            chessBoard = new ChessBoard();
        }

        [Test]
        public void Has_MaxBoardWidth_of_8()
        {
            Assert.AreEqual(8, ChessBoard.MaxBoardWidth);
        }

        [Test]
        public void Has_MaxBoardHeight_of_8()
        {
            Assert.AreEqual(8, ChessBoard.MaxBoardHeight);
        }

        // Simplifying by doing boundary checks
        [TestCase(0, 0, true)]
        [TestCase(0, 7, true)]
        [TestCase(7, 0, true)]
        [TestCase(7, 7, true)]
        [TestCase(8, 0, false)]
        [TestCase(0, 8, false)]
        [TestCase(-1, 0, false)]
        [TestCase(0, -1, false)]
        public void IsLegalBoardPosition_True(int xPos, int yPos, bool expected)
        {
            var isValidPosition = ChessBoard.IsLegalBoardPosition(xPos, yPos);
            Assert.AreEqual(expected, isValidPosition);
        }

        [Test]
        public void Avoids_Duplicate_Positioning()
        {
            Pawn firstPawn = new Pawn(PieceColor.Black);
            Pawn secondPawn = new Pawn(PieceColor.Black);
            chessBoard.Add(firstPawn, 6, 3);
            chessBoard.Add(secondPawn, 6, 3);
            Assert.AreEqual(6, firstPawn.XCoordinate);
            Assert.AreEqual(3, firstPawn.YCoordinate);
            Assert.AreEqual(-1, secondPawn.XCoordinate);
            Assert.AreEqual(-1, secondPawn.YCoordinate);
        }

        [Test]
        public void GetPiece_Remove_Pawn()
        {
            Pawn pawn = new Pawn(PieceColor.White);
            chessBoard.Add(pawn, 4, 1);
            Assert.AreEqual(pawn, chessBoard.GetPiece(4, 1));
            chessBoard.Remove(4, 1);
            Assert.AreEqual(null, chessBoard.GetPiece(4, 1));
        }

        [Test]
        public void GetPiece_Remove_Invalid_Pawn()
        {
            Assert.AreEqual(null, chessBoard.GetPiece(-1, 0));
            chessBoard.Remove(0, -1);
            Assert.AreEqual(null, chessBoard.GetPiece(0, -1));
        }

        [Test]
        public void Limits_The_Number_Of_Pawns()
        {
            // place black pawns on the starting row and any 
            // overspill on the row below
            for (int i = 0; i < 10; i++)
            {
                Pawn pawn = new Pawn(PieceColor.Black);
                int column = i % ChessBoard.MaxBoardWidth;
                int row = 6 - i / ChessBoard.MaxBoardWidth;
                chessBoard.Add(pawn, column, row);
                if (row == 6)
                {
                    Assert.AreEqual(column, pawn.XCoordinate);
                    Assert.AreEqual(row, pawn.YCoordinate);
                }
                else
                {
                    Assert.AreEqual(-1, pawn.XCoordinate);
                    Assert.AreEqual(-1, pawn.YCoordinate);
                }
            }
        }

        [Test]
        public void Limit_Updated_Removed_Pawn()
        {
            for (int i = 0; i < 8; i++)
            {
                Pawn p = new Pawn(PieceColor.White);
                chessBoard.Add(p, i, 1);
                Assert.AreEqual(i, p.XCoordinate);
                Assert.AreEqual(1, p.YCoordinate);
            }
            Pawn pawn = new Pawn(PieceColor.White);
            chessBoard.Add(pawn, 0, 2);
            Assert.AreEqual(-1, pawn.XCoordinate);
            Assert.AreEqual(-1, pawn.YCoordinate);
            chessBoard.Remove(0, 1);
            chessBoard.Add(pawn, 0, 2);
            Assert.AreEqual(0, pawn.XCoordinate);
            Assert.AreEqual(2, pawn.YCoordinate);
        }
    }
}
